File: java/Hash Maps and Sets/LongestChainOfConsecutiveNumbers.java
import java.util.HashSet;
import java.util.Set;

public class LongestChainOfConsecutiveNumbers {
    public int longestChainOfConsecutiveNumbers(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        Set<Integer> numSet = new HashSet<>();
        for (int num : nums) {
            numSet.add(num);
        }
        int longestChain = 0;
        for (int num : numSet) {
            // If the current number is the smallest number in its chain, search for
            // the length of its chain.
            if (!numSet.contains(num - 1)) {
                int currentNum = num;
                int currentChain = 1;
                // Continue to find the next consecutive numbers in the chain.
                while (numSet.contains(currentNum + 1)) {
                    currentNum++;
                    currentChain++;
                }
                longestChain = Math.max(longestChain, currentChain);
            }
        }
        return longestChain;
    }
}

File: go/Hash Maps and Sets/longest_chain_of_consecutive_numbers.go
func longestChainOfConsecutiveNumbers(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	numSet := make(map[int]bool)
	for _, num := range nums {
		numSet[num] = true
	}
	longestChain := 0
	for num := range numSet {
		// If the current number is the smallest number in its chain, search for
		// the length of its chain.
		if !numSet[num-1] {
			currentNum := num
			currentChain := 1
			// Continue to find the next consecutive numbers in the chain.
			for numSet[currentNum+1] {
				currentNum++
				currentChain++
			}
			if currentChain > longestChain {
				longestChain = currentChain
			}
		}
	}
	return longestChain
}
